include/slate/BaseMatrix.hh

```diff
@@ -1283,34 +1284,43 @@ void swap(BaseMatrix<scalar_t>& A, BaseMatrix<scalar_t>& B)
 ///
 template <typename scalar_t>
 void BaseMatrix<scalar_t>::gridinfo(
-    GridOrder* order, int* nprow, int* npcol, int* myrow, int* mycol ) const
-{
-    if (nprow_ > 0) {
-        *order = order_;
-        *nprow = nprow_;
-        *npcol = npcol_;
-        if (order_ == GridOrder::Col) {
-            *myrow = mpi_rank_ % nprow_;
-            *mycol = mpi_rank_ / nprow_;
-        }
-        else {
-            *myrow = mpi_rank_ / npcol_;
-            *mycol = mpi_rank_ % npcol_;
-        }
-    }
-    else {
+    GridOrder* order, int* nprow, int* npcol, int* myrow, int* mycol )
+{
+    // If grid order is unknown, attempt to detect it
+    if (order_ == GridOrder::Unknown) {
         int mpi_size;
-        slate_mpi_call(
-            MPI_Comm_size(mpiComm(), &mpi_size));
-        if (mpi_size == 1) {
-            *order = GridOrder::Col;
-            *nprow = *npcol = 1;
-            *myrow = *mycol = 0;
-        }
-        else {
-            *order = GridOrder::Unknown;
-            *nprow = *npcol = *myrow = *mycol = -1;
+        slate_mpi_call(MPI_Comm_size(mpiComm(), &mpi_size));
+        // When matrix is empty, storage_ may be null
+        if (mt_ != 0 && nt_ != 0) {
+            int64_t nprow_i64, npcol_i64;
+            func::is_2d_cyclic_grid(mt(), nt(), tileRankFunc(), order_, nprow_i64, npcol_i64);
+            nprow_ = nprow_i64;
+            npcol_ = npcol_i64;
+
+            // Detected grid doesn't match process distribution
+            // Label it as unknown since some processes won't have a row/column
+            if (order_ != GridOrder::Unknown && mpi_size < nprow_*npcol_) {
+                order_ = GridOrder::Unknown;
+                nprow_ = npcol_ = -1;
+            }
```

Comment:
This appears to say it detected a 2D grid that is bigger than the MPI process grid. When would this happen, can you give an example?

Reply:
We don't do any validation of the distribution functions in the lambda constructor, so users can accidentally give us things like `process_2d_grid(ColMajor, mpi_size, mpi_size)`.  It's maybe an unnecessary guard since we're probably going to be in a bad state anyways.  But, I figure that we get a little extra robustness for just a couple extra ops.

Reply:
Ok.